lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp

```diff
@@ -383,50 +700,123 @@ void IMConstPropPass::markBlockExecutable(Block *block) {
       markRegResetOp(regReset);
     else if (auto mem = dyn_cast<MemOp>(op))
       markMemOp(mem);
+    else if (auto subindex = dyn_cast<SubindexOp>(op))
+      markSubindexOp(subindex);
+    else if (auto subfield = dyn_cast<SubfieldOp>(op))
+      markSubfieldOp(subfield);
   }
 }
 
 void IMConstPropPass::markWireOrUnresetableRegOp(Operation *wireOrReg) {
-  // If the wire/reg has a non-ground type, then it is too complex for us to
-  // handle, mark it as overdefined.
-  // TODO: Eventually add a field-sensitive model.
+  // If the wire/reg has a non-ground type, then it is too complex for us
+  // to handle, mark it as overdefined.
   auto resultValue = wireOrReg->getResult(0);
-  if (!resultValue.getType().cast<FIRRTLType>().getPassiveType().isGround())
+  if (!isTrackableType(resultValue.getType()))
     return markOverdefined(resultValue);
 
-  // Otherwise, this starts out as InvalidValue and is upgraded by connects.
-  mergeLatticeValue(resultValue, InvalidValueAttr::get(resultValue.getType()));
+  // Otherwise, this starts out as InvalidValue and is upgraded by
+  // connects.
+  auto &destTypes = getLeafGroundTypes(resultValue.getType());
+  for (unsigned i = 0, e = destTypes.size(); i < e; i++) {
+    auto destType = destTypes[i].cast<FIRRTLType>();
+    mergeLatticeValue({resultValue, i, /*isKnownRoot=*/true},
+                      InvalidValueAttr::get(destType));
+  }
 }
 
 void IMConstPropPass::markRegResetOp(RegResetOp regReset) {
-  // If the reg has a non-ground type, then it is too complex for us to handle,
-  // mark it as overdefined.
-  // TODO: Eventually add a field-sensitive model.
-  if (!regReset.getType().getPassiveType().isGround())
+  // If the reg has a passive type, then it is too complex for us to
+  // handle, mark it as overdefined.
+  if (!isTrackableType(regReset.getType()))
     return markOverdefined(regReset);
 
   // The reset value may be known - if so, merge it in.
-  auto srcValue = getExtendedLatticeValue(regReset.resetValue(),
-                                          regReset.getType().cast<FIRRTLType>(),
+  auto &destTypes = getLeafGroundTypes(regReset.getType());
+
+  // Iterate over each ground type and merge lattice values separatetly.
+  for (unsigned i = 0, e = destTypes.size(); i < e; i++) {
```

Comment:
Nit: the convention in CIRCT tends to be towards always using preincrement, even though it doesn't matter here. Ditto for other uses here.

Reply:
Thanks!